elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/impl/ElasticsearchHSQueryImpl.java

```diff
@@ -553,6 +586,78 @@ private JsonElement getFieldValue(JsonObject parent, String projectedField) {
 		}
 	}
 
+	@Override
+	protected void extractFacetResults() {
+		SearchResult searchResult = getSearchResult();
+		JsonElement aggregationsElement = searchResult.getJsonObject().get( "aggregations" );
+		if ( aggregationsElement == null ) {
+			return;
+		}
+		JsonObject aggregations = aggregationsElement.getAsJsonObject();
+
+		Map<String, List<Facet>> results = new HashMap<>();
+		for ( FacetingRequest facetRequest : getFacetManager().getFacetRequests().values() ) {
+			List<Facet> facets;
+			if ( facetRequest instanceof DiscreteFacetRequest ) {
+				facets = updateStringFacets( aggregations, (DiscreteFacetRequest) facetRequest );
+				// Discrete facets are sorted by Elasticsearch
+			}
+			else {
+				facets = updateRangeFacets( aggregations, (RangeFacetRequest<?>) facetRequest );
+				if ( !FacetSortOrder.RANGE_DEFINITION_ORDER.equals( facetRequest.getSort() ) ) {
+					Collections.sort( facets, facetComparators.get( facetRequest.getSort() ) );
+				}
+			}
+
+			results.put( facetRequest.getFacetingName(), facets );
+		}
+		getFacetManager().setFacetResults( results );
+	}
+
+	private List<Facet> updateRangeFacets(JsonObject aggregations, RangeFacetRequest<?> facetRequest) {
+		if ( !ReflectionHelper.isIntegerType( facetRequest.getFacetValueType() )
+				&& !Date.class.isAssignableFrom( facetRequest.getFacetValueType() )
+				&& !ReflectionHelper.isFloatingPointType( facetRequest.getFacetValueType() ) ) {
+			throw LOG.unsupportedFacetRangeParameter( facetRequest.getFacetValueType().getName() );
+		}
+
+		ArrayList<Facet> facets = new ArrayList<>();
+		for ( FacetRange<?> facetRange : facetRequest.getFacetRangeList() ) {
+			JsonElement aggregation = aggregations.get( facetRequest.getFacetingName() + "-" + facetRange.hashCode() );
+			if ( aggregation == null ) {
+				continue;
+			}
+			int docCount = aggregation.getAsJsonObject().get( "doc_count" ).getAsInt();
+			if ( docCount == 0 && !facetRequest.hasZeroCountsIncluded() ) {
+				continue;
+			}
+			facets.add( facetRequest.createFacet( facetRange.getRangeString(), docCount ) );
+		}
+		return facets;
+	}
+
+	private List<Facet> updateStringFacets(JsonObject aggregations, DiscreteFacetRequest facetRequest) {
+		JsonElement aggregation = aggregations.get( facetRequest.getFacetingName() );
+		if ( aggregation == null ) {
+			return Collections.emptyList();
+		}
+		// deal with nested aggregation for nested documents
+		if ( facetRequest.getFieldName().contains( "." ) ) {
```

Comment:
Yes, my point was just that I don't think it's something we should fix in this PR.

I can take a shot at it if I'm not stepping on your toes?

Created https://hibernate.atlassian.net/browse/HSEARCH-2097 so that we don't forget about it.